### DISCUSSION TITLE
Support Passing Trace IDs in Requests

### DISCUSSION THREAD

**Discussion by theory** [ory/hydra#3414] [Q&A]
### Preflight checklist

- [X] I could not find a solution in the existing issues, docs, nor discussions.
- [X] I agree to follow this project's [Code of Conduct](https://github.com/ory/hydra/blob/master/CODE_OF_CONDUCT.md).
- [X] I have read and am following this repository's [Contribution Guidelines](https://github.com/ory/hydra/blob/master/CONTRIBUTING.md).
- [ ] This issue affects my [Ory Network](https://www.ory.sh/) project.
- [X] I have joined the [Ory Community Slack](https://slack.ory.sh).
- [ ] I am signed up to the [Ory Security Patch Newsletter](https://ory.us10.list-manage.com/subscribe?u=ffb1a878e4ec6c0ed312a3480&id=f605a41b53).

### Describe your problem

The tracing support in Hydra works great; we have it wired up so that Datadog collects traces (we're on 1.11.10). It's great.

However we noticed that although our delegate service (which handles the actual authentication bits) makes calls to Hydra, it does not appear in Hyda's traces. This is because we currently have no way to inject the trace ID into requests we make to Hydra. For example, here's how to inject the trace ID into a standard Go `http.Request` using [opentracing.Inject](https://pkg.go.dev/github.com/opentracing/opentracing-go@v1.2.0#readme-serializing-to-the-wire):

``` go
opentracing.GlobalTracer().Inject(
	span.Context(),
	opentracing.HTTPHeaders,
	opentracing.HTTPHeadersCarrier(req.Header),
)
```

This is do-able because we have direct access to the headers submitted in the request. Alas, the [Hydra Go client](https://pkg.go.dev/github.com/ory/hydra-client-go@v1.11.8) does not provide an interface to allow access to request headers. If there was an interface to allow for adding custom headers for every request, then just about any tracing library would be able to inject the necessary trace ID to ensure that all services show up in the traces.

### Describe your ideal solution

This is a request for that interface, but also, if something needs to be done in Hydra itself to make sure that such IDs in requests it handles are propagated to the traces it emits, that would be the necessary other half. I suspect the most tracers enable this through middleware.

### Workarounds or alternatives

No known alternatives; I poked around the  [Hydra Go client](https://pkg.go.dev/github.com/ory/hydra-client-go@v1.11.8) source but could not find any public access to the http.Request; `func (c *APIClient) prepareRequest` is  obviously private.

### Version

1.11.10

### Additional Context

[Slack post](https://ory-community.slack.com/archives/C012RBW0F18/p1672859952908449) and [thread](https://ory-community.slack.com/archives/C012RBW0F18/p1672860116262919).

**Comment by aeneasr** [comment]
You're right, I also could not find an easy way to solve this. One way of doing it would probably to create a custom transport (or maybe use a transport from otel?) that can inject the appropriate HTTP header. Then, you can customize the http client with:

```
import ory "github.com/ory/client-go"

cfg := ory.NewConfiguration()
cfg.HTTPClient.Transport = NewInjectTracingTransport()
```

I think that should do the trick.

**Reply by theory** [reply]
That make sense. I'd like to request that an example be added to the docs for the SDK.
